### PR TITLE
unify moe implementation for llama4 and deepseek_v3

### PR DESCRIPTION
**Pull request by tianyu-l**
Given the complexity of MoE and EP modules
This PR
1. creates `torchtitan/models/moe.py` as the central moe implementation (this is similar to why we have `torchtitan/models/attention.py`)
2. creates `torchtitan/distributed/expert_parallel.py` as the central EP implementation
3. rename `torchtitan/distributed/pipeline.py` -> `torchtitan/distributed/pipeline_parallel.py` to be consistent with EP
4. apply temporary fix by @rakkit https://github.com/pytorch/torchtitan/issues/1467 before the memory leak issue with AC + PT-D all_to_all_single_autograd is fixed (cc @soulitzer)

### CODE REVIEW

File: torchtitan/distributed/expert_parallel.py [hunk 1]
```diff
@@ -11,7 +11,8 @@
 import torch
 import torch.distributed as dist
 import torch.nn as nn
-from torch.distributed._functional_collectives import all_to_all_single_autograd
+
+# from torch.distributed._functional_collectives import all_to_all_single_autograd
```
Comment (danielvegamyhre) [inline torchtitan/distributed/expert_parallel.py]: nit: remove commented code

File: torchtitan/distributed/expert_parallel.py [hunk 2]
```diff
@@ -24,6 +25,40 @@
 from torch.distributed.tensor.placement_types import Placement
 
 
+# TODO: there is memory leak issue with AC + PT-D all_to_all_single_autograd
+# This is a temporary fix by @rakkit https://github.com/pytorch/torchtitan/issues/1467
+class _A2A(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, out_splits, in_splits, group):
+        if isinstance(out_splits, torch.Tensor):
+            out_splits = out_splits.tolist()
```
Comment (danielvegamyhre) [inline torchtitan/distributed/expert_parallel.py:33]: won't tolist() cause d2h sync? is this okay / intentional in this case?

**Review by danielvegamyhre — APPROVED** [review]
lgtm, left a couple of minor comments/questions.

File: torchtitan/distributed/expert_parallel.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/distributed/expert_parallel.py]: this is intentional -- we should restore this implementation after bug is fixed. I reorganized the code a bit to make it clearer.

File: torchtitan/distributed/expert_parallel.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/distributed/expert_parallel.py:33]: It will. This is a temporary fix, but currently in EP there are multiple places with d2h sync. I'm working on another implementation to kill them.

**Review by ruisizhang123 — APPROVED** [review]
LGTM, thank you for the refactor.

File: torchtitan/models/moe.py [hunk 3]
```diff
@@ -4,13 +4,32 @@
...
 import torch.nn.functional as F
 from torch import nn
 
-from ..infra.expert_parallel import expert_parallel
+from torchtitan.distributed.expert_parallel import expert_parallel
+
+
+@dataclass
+class MoEArgs:
+    moe_enabled: bool = True
```
Comment (wwwjn) [inline torchtitan/models/moe.py]: Why we need to have `moe_enabled` in MoEArgs?

Comment (wwwjn) [reply torchtitan/models/moe.py]: I didn't see anywhere this parameter is `false`

**Comment by rakkit** [comment]
(sry, I am leaving for vacation and lazy to open PR). 
Since you are refactoring MOE, here is another trick or bug we found.   When we use DP2EP, the gradient reduction denominator (of EP) is always smaller than other modules(e.g. attention or embedding) so the actual gradient of EP is always higher. See logs.

image

we fixed this by adding a `loss_average_denominator` in `parallel_dims.py` and forcing EP's reduce denominator to 
```
            transformer_block.feed_forward.experts.set_reduce_scatter_divide_factor(
                loss_average_denominator,
            )
 ```  
 in `apply_fsdp`, you can also check full code [here](https://github.com/rakkit/torchtitan/blob/trustllm-0.1.5/torchtitan/models/MoEllama/infra/parallelize.py#L473)

where `loss_average_denominator = dp_replicate * dp_shard *cp`  [see  here](https://github.com/rakkit/torchtitan/blob/trustllm-0.1.5/torchtitan/distributed/parallel_dims.py#L223)

 ----

 We also have a better version of the bias update that only needs one reduce, you can check the code [here](https://github.com/rakkit/torchtitan/blob/trustllm-0.1.5/torchtitan/models/MoEllama/infra/build_optimizer.py).     A tricks here does not affect to Bias update, but we need to know, once we have activation checkpoint.
 ```
         if self.load_balance_coeff is not None:
            with torch.no_grad():
                self.tokens_per_expert.add_(num_tokens_per_expert)
 ```
It will be called more than once, so the actual stats value of `num_tokens_per_expert` will be 2x.

File: torchtitan/models/moe.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/moe.py]: makes sense, removed

**Comment by garrett361** [comment]
Nice @rakkit, we found the same issue with the ep grads being off by a factor. I was finding that [set_reduce_scatter_divide_factor errored](https://github.com/pytorch/pytorch/issues/155223#issuecomment-2956670292)  when using an mp policy, though. 

Surprised you didn't hit that? Think I saw you're on `torch==2.6` in another comment elsewhere

**Comment by rakkit** [comment]
lol @garrett361  thanks for the info.  I did not see the issue on both Torch 2.6 and 2.7.1.

To clarify I only test the default mp set(mixed_precision_param=bf16 and mixed_precision_reduce=fp32)
